### ISSUE TITLE
Cancelling a task causes cancelling of every forked task?

### ISSUE THREAD

**Issue by michalczaplinski** [yelouafi/redux-saga#478]
Hi, I'm still new to `redux-saga` and have the following problem: 

I have a simple list of items in my UI (the "bookmarks" in the code below). When I delete one of them, I want to show an "undo" button and then wait 5 seconds for the the user's action. If the user clicks on the undo button, the we dispatch the `types.CANCEL_DELETE_BOOKMARK` action, otherwise, the saga calls the API. 

The issue I seem to have is that if I dispatch multiple `types.DELETE_BOOKMARK` in a quick succession, then a subsequent dispatching of `types.CANCEL_DELETE_BOOKMARK` cancels every `performDelete`, not just the one in the forked task.

```javascript
function* performDelete(id) {
  try {
    yield call(delay, 5000);
    yield call(api.deleteBookmark, id);

    yield put({type: types.DELETE_BOOKMARK_SUCCESS, id});
  } catch(error) {
    yield put({type: types.DELETE_BOOKMARK_ERROR, error});
  }
}


function* deleteBookmark(act) {
  const { id } = act;
  yield put({type: types.REMOVE_BOOKMARK_FROM_UI, id});
  yield put({type: types.SHOW_UNDO, id});

  const { response, cancelDeleting } = yield race({
    delete: call(performDelete, id),
    cancelDeleting: take([types.CANCEL_DELETE_BOOKMARK, types.DELETE_BOOKMARK_ERROR])
  });

  if (cancelDeleting) {
    yield put({type: types.READD_BOOKMARK_TO_UI, id});
  }
  yield put({type: types.HIDE_UNDO, id});
}


export default function* rootSaga() {
  yield* takeEvery(types.DELETE_BOOKMARK, deleteBookmark)
}
```

I have also tried to fork and cancel the `performDelete` task, without the use of `race`, but that results in the same behavior.

**Comment by jrvidal** [comment]
If you fire several `DELETE_BOOKMARK` in succession, won't a single `DELETE_BOOKMARK_ERROR` cancel all your races? Also, shouldn't `deleteBookmark` take an `action` parameter instead of an `id`?

**Comment by michalczaplinski** [comment]
Yes you are right, the generator takes the `action` as the parameter, of course. That was just a copy-paste mistake and not part of the issue I have.

> If you fire several DELETE_BOOKMARK in succession, won't a single DELETE_BOOKMARK_ERROR cancel all your races?

No, the `DELETE_BOOKMARK_ERROR` as well as CANCEL_DELETE_BOOKMARK (which is what I trigger from the UI) should both only cancel the race in **one** of the forked tasks. That is what I intended and how I understood `takeEvery` to work at least... 😕

**Comment by jrvidal** [comment]
> should both only cancel the race in one of the forked tasks

I don't think that is the case. All your concurrent `race`s are blocked on a `take` effect that will "win" on a `DELETE_BOOKMARK_ERROR` action. This action is "global": every `take` blocked on it will resume after receiving it. `takeEvery` does not play any role here.

**Comment by michalczaplinski** [comment]
Okay, I see. Could you explain 
> All your concurrent races are blocked on a take effect that will "win" on a DELETE_BOOKMARK_ERROR action

in a bit more detail? It's not clear to me why **all** of the races are blocked. I was under the impression that in my example every `DELETE_BOOKMARK` action forks a new task (thanks to `takeEvery`) and then inside the `deleteBookmark` task there is a race, so that would mean there is a race for each forked task. Could you show an example that would explain this better for me?

Also, the `DELETE_BOOKMARK_ERROR` action is not the issue, because it is not triggered in the example I have. I mean, if I remove the `catch` clause in the `performDelete` and remove `types.DELETE_BOOKMARK_ERROR` from the `take` call in `deleteBookmark`, I am left with the same error...

Do you see what I mean?

**Comment by jrvidal** [comment]
You're right the `DELETE_BOOKMARK_ERROR` action is not the issue. I noticed it first for that action for some reason, but the case is the same for `CANCEL_DELETE_BOOKMARK`.

What you probably need is something like:
```javascript
const { response, cancelDeleting } = yield race({
  delete: call(performDelete, id),
  cancelDeleting: take(action => (
    (action.type === types.DELETE_BOOKMARK_ERROR || action.type === types.CANCEL_DELETE_BOOKMARK) &&
    action.id === id
  ))
});
```

Otherwise `CANCEL_DELETE_BOOKMARK` affects all your races. There is nothing in your `take` referring to the particular bookmark you are dealing with.

**Comment by michalczaplinski** [comment]
That makes sense! Thanks a lot! I somehow had a completely wrong mental model and just assumed that the actions are somehow scoped inside the tasks. Perhaps this could be made more clear in the docs.
